### PR TITLE
[new release] lab, gitlab, gitlab-unix and gitlab-jsoo (0.1.0)

### PR DESCRIPTION
**Pull request by tmcgilchrist**
GitLab cli

- Project page: <a href="https://github.com/tmcgilchrist/ocaml-gitlab">https://github.com/tmcgilchrist/ocaml-gitlab</a>
- Documentation: <a href="https://tmcgilchrist.github.io/ocaml-gitlab/">https://tmcgilchrist.github.io/ocaml-gitlab/</a>

## Added

- Initial release of gitlab bindings. Supports Personal and Project Access Tokens, and selected
  parts of the Project, User, Events and Group APIs.

### CODE REVIEW

File: packages/lab/lab.0.1.0/opam [hunk 1]
```diff
@@ -0,0 +1,45 @@
...
+doc: "https://tmcgilchrist.github.io/ocaml-gitlab/"
+bug-reports: "https://github.com/tmcgilchrist/ocaml-gitlab/issues"
+depends: [
+  "dune" {>= "2.9"}
+  "ocaml" {>= "4.08.0"}
+  "mdx" {with-test}
+  "cmdliner" {>= "0.9.8"}
+  "gitlab-unix" {= version}
+  "otoml" {>= "0.9.0"}
+  "odoc" {with-doc}
```
Comment (mseri) [inline packages/lab/lab.0.1.0/opam:18]:
```suggestion
  "alcotest-lwt" {with-test}
  "odoc" {with-doc}
```

alcotest-lwt is required by `lab` tests.

File: packages/gitlab/gitlab.0.1.0/opam [hunk 2]
```diff
@@ -0,0 +1,50 @@
...
+bug-reports: "https://github.com/tmcgilchrist/ocaml-gitlab/issues"
+depends: [
+  "ocaml" {>= "4.08.0"}
+  "dune" {>= "2.9" & >= "1.10"}
+  "uri" {>= "1.9.0"}
+  "cohttp-lwt" {>= "4.0"}
+  "atdgen" {>= "2.0.0"}
+  "yojson" {>= "1.7.0"}
+  "stringext"
+  "odoc" {with-doc}
```
Comment (mseri) [inline packages/gitlab/gitlab.0.1.0/opam:23]:
```suggestion
  "alcotest-lwt" {with-test}
  "odoc" {with-doc}
```

File: packages/gitlab-unix/gitlab-unix.0.1.0/opam [hunk 3]
```diff
@@ -0,0 +1,54 @@
...
+  [
+    "dune"
+    "build"
+    "-p"
+    name
+    "-j"
+    jobs
+    "--promote-install-files=false"
+    "@install"
+    "@runtest" {with-test}
```
Comment (mseri) [inline packages/gitlab-unix/gitlab-unix.0.1.0/opam:40]:
```suggestion
#    "@runtest" {with-test}
```
Requires an external dependency:
```
- /bin/sh: 1: jd: not found
```

File: packages/gitlab-jsoo/gitlab-jsoo.0.1.0/opam [hunk 4]
```diff
@@ -0,0 +1,47 @@
...
+doc: "https://tmcgilchrist.github.io/ocaml-gitlab/"
+bug-reports: "https://github.com/tmcgilchrist/ocaml-gitlab/issues"
+depends: [
+  "dune" {>= "2.9"}
+  "ocaml" {>= "4.08.0"}
+  "gitlab" {= version}
+  "cohttp" {>= "4.0"}
+  "cohttp-lwt-jsoo" {>= "4.0"}
+  "js_of_ocaml-lwt" {>= "3.4.0"}
+  "odoc" {with-doc}
```
Comment (mseri) [inline packages/gitlab-jsoo/gitlab-jsoo.0.1.0/opam:20]:
```suggestion
  "mdx" {with-test}
  "alcotest-lwt" {with-test}
  "odoc" {with-doc}
```
The tests are showing missing mdx and missing alcotest-lwt failures

File: packages/gitlab-unix/gitlab-unix.0.1.0/opam [hunk 5]
```diff
@@ -0,0 +1,54 @@
...
+  "cohttp" {>= "4.0"}
+  "cohttp-lwt-unix" {>= "4.0"}
+  "tls" {>= "0.11.0"}
+  "lwt" {>= "2.4.4"}
+  "stringext"
+  "alcotest" {with-test}
+  "alcotest-lwt" {with-test}
+  "qcheck" {with-test}
+  "qcheck-alcotest" {with-test}
+  "base-unix"
```
Comment (mseri) [inline packages/gitlab-unix/gitlab-unix.0.1.0/opam:26]:
```suggestion
  "mdx" {with-test}
  "base-unix"
```

File: packages/lab/lab.0.1.0/opam [hunk 6]
```diff
@@ -0,0 +1,46 @@
...
+doc: "https://tmcgilchrist.github.io/ocaml-gitlab/"
+bug-reports: "https://github.com/tmcgilchrist/ocaml-gitlab/issues"
+depends: [
+  "dune" {>= "2.9"}
+  "ocaml" {>= "4.08.0"}
+  "mdx" {with-test}
+  "cmdliner" {>= "0.9.8"}
+  "gitlab-unix" {= version}
+  "otoml" {>= "0.9.0"}
+  "alcotest-lwt" {with-test}
```
Comment (kit-ty-kate) [inline packages/lab/lab.0.1.0/opam:18]:
```suggestion
  "alcotest-lwt" {with-test}
  "qcheck-alcotest" {with-test}
  "lab" {= version & with-test}
```

File: packages/gitlab/gitlab.0.1.0/opam [hunk 7]
```diff
@@ -0,0 +1,52 @@
...
+depends: [
+  "ocaml" {>= "4.08.0"}
+  "dune" {>= "2.9" & >= "1.10"}
+  "uri" {>= "1.9.0"}
+  "cohttp-lwt" {>= "4.0"}
+  "atdgen" {>= "2.0.0"}
+  "yojson" {>= "1.7.0"}
+  "stringext"
+  "mdx" {with-test}
+  "alcotest-lwt" {with-test}
```
Comment (kit-ty-kate) [inline packages/gitlab/gitlab.0.1.0/opam:24]:
```suggestion
  "alcotest-lwt" {with-test}
  "qcheck-alcotest" {with-test}
  "lab" {= version & with-test}
```

File: packages/gitlab-jsoo/gitlab-jsoo.0.1.0/opam [hunk 8]
```diff
@@ -0,0 +1,49 @@
...
+bug-reports: "https://github.com/tmcgilchrist/ocaml-gitlab/issues"
+depends: [
+  "dune" {>= "2.9"}
+  "ocaml" {>= "4.08.0"}
+  "gitlab" {= version}
+  "cohttp" {>= "4.0"}
+  "cohttp-lwt-jsoo" {>= "4.0"}
+  "js_of_ocaml-lwt" {>= "3.4.0"}
+  "mdx" {with-test}
+  "alcotest-lwt" {with-test}
```
Comment (kit-ty-kate) [inline packages/gitlab-jsoo/gitlab-jsoo.0.1.0/opam:21]:
```suggestion
  "alcotest-lwt" {with-test}
  "qcheck-alcotest" {with-test}
  "lab" {= version & with-test}
```

**Review by kit-ty-kate — COMMENTED** [review]
```

#=== ERROR while compiling gitlab.0.1.0 =======================================#
# context              2.0.10 | linux/x86_64 | ocaml-base-compiler.4.08.1 | pinned(https://github.com/tmcgilchrist/ocaml-gitlab/releases/download/0.1.0/lab-0.1.0.tbz)
# path                 ~/.opam/4.08/.opam-switch/build/gitlab.0.1.0
# command              ~/.opam/opam-init/hooks/sandbox.sh build dune build -p gitlab -j 31 --promote-install-files=false @install @runtest
# exit-code            1
# env-file             ~/.opam/log/gitlab-2958-5f1023.env
# output-file          ~/.opam/log/gitlab-2958-5f1023.out
### output ###
# File "test/dune", line 3, characters 47-62:
# 3 |  (libraries gitlab gitlab-unix alcotest qcheck qcheck-alcotest alcotest-lwt)
#                                                    ^^^^^^^^^^^^^^^
# Error: Library "qcheck-alcotest" not found.
# Hint: try:
#   dune external-lib-deps --missing --no-config --root . --ignore-promoted-rules --default-target @install --always-show-command-line --promote-install-files false --promote-install-files --release --only-packages gitlab -p gitlab --profile release -j 31 @install @runtest
# File "cli/dune", line 9, characters 11-14:
# 9 |  (packages lab))
#                ^^^
# Error: Package lab does not exist
```

**Comment by tmcgilchrist** [comment]
Thank you for the edits @kit-ty-kate @mseri 
I'm re-tagging the source with various fixes.

**Comment by mseri** [comment]
This seems to appear as error in all packages:
```
# File "cli/dune", line 7, characters 0-37:
# 7 | (mdx
# 8 |  (files lab.md)
# 9 |  (packages lab))
# Error: Program ocaml-mdx not found in the tree or in PATH
#  (context: default)
# Hint: opam install mdx
# File "test/dune", line 28, characters 0-79:
# 28 | (rule
# 29 |  (with-stdout-to
# 30 |   user_short_output.json
# 31 |   (run ./diff.exe user_short)))
# Error: No rule found for test/diff.exe
```

Also there are some failures related to attempting network connections (which are not allowed in the sandbox). I would suggest to disable the tests in that case

**Comment by mseri** [comment]
The other issue is
```
#=== ERROR while compiling gitlab-unix.0.1.0 ==================================#
# context              2.0.10 | linux/x86_64 | ocaml-base-compiler.4.08.1 | pinned(https://github.com/tmcgilchrist/ocaml-gitlab/releases/download/0.1.0/lab-0.1.0.tbz)
# path                 ~/.opam/4.08/.opam-switch/build/gitlab-unix.0.1.0
# command              ~/.opam/opam-init/hooks/sandbox.sh build dune build -p gitlab-unix -j 31 --promote-install-files=false @install @runtest
# exit-code            1
# env-file             ~/.opam/log/gitlab-unix-687-5f1023.env
# output-file          ~/.opam/log/gitlab-unix-687-5f1023.out
### output ###
# File "cli/dune", line 7, characters 0-37:
# 7 | (mdx
# 8 |  (files lab.md)
# 9 |  (packages lab))
# Error: Program ocaml-mdx not found in the tree or in PATH
#  (context: default)
# Hint: opam install mdx
#         diff test/branches_output.json (exit 1)
# (cd _build/default/test && ./diff.exe branches) > _build/default/test/branches_output.json
# diff: internal error, uncaught exception:
#       Sys_error("cases/branches/event.json: No such file or directory")
#       
#         diff test/commit_statuses_output.json (exit 1)
# (cd _build/default/test && ./diff.exe commit_statuses) > _build/default/test/commit_statuses_output.json
# diff: internal error, uncaught exception:
#       Sys_error("cases/commit_statuses/event.json: No such file or directory")
#       
#         diff test/events_output.json (exit 1)
# (cd _build/default/test && ./diff.exe events) > _build/default/test/events_output.json
# diff: internal error, uncaught exception:
#       Sys_error("cases/events/event.json: No such file or directory")
#       
#         diff test/merge_requests_output.json (exit 1)
# (cd _build/default/test && ./diff.exe merge_requests) > _build/default/test/merge_requests_output.json
# diff: internal error, uncaught exception:
#       Sys_error("cases/merge_requests/event.json: No such file or directory")
#       
#         diff test/milestones_output.json (exit 1)
# (cd _build/default/test && ./diff.exe milestones) > _build/default/test/milestones_output.json
# diff: internal error, uncaught exception:
#       Sys_error("cases/milestones/event.json: No such file or directory")
#       
#         diff test/user_short_output.json (exit 1)
# (cd _build/default/test && ./diff.exe user_short) > _build/default/test/user_short_output.json
# diff: internal error, uncaught exception:
#       Sys_error("cases/user_short/event.json: No such file or directory")
#       
#         diff test/webhooks_output.json (exit 1)
# (cd _build/default/test && ./diff.exe webhooks) > _build/default/test/webhooks_output.json
# diff: internal error, uncaught exception:
#       Sys_error("cases/webhooks/event.json: No such file or directory")
#       
```

File: packages/gitlab-jsoo/gitlab-jsoo.0.1.0/opam (hunk 4 above)
Comment (mseri) [inline packages/gitlab-jsoo/gitlab-jsoo.0.1.0/opam:20]:
```suggestion
  "mdx" {with-test}
  "odoc" {with-doc}
```

File: packages/gitlab-unix/gitlab-unix.0.1.0/opam [hunk 9]
```diff
@@ -0,0 +1,52 @@
...
+  "gitlab" {= version}
+  "cohttp" {>= "4.0"}
+  "cohttp-lwt-unix" {>= "4.0"}
+  "tls" {>= "0.11.0"}
+  "lwt" {>= "2.4.4"}
+  "stringext"
+  "base-unix"
+  "alcotest" {with-test}
+  "alcotest-lwt" {with-test}
+  "odoc" {with-doc}
```
Comment (mseri) [inline packages/gitlab-unix/gitlab-unix.0.1.0/opam:25]:
```suggestion
  "mdx" {with-test}
  "odoc" {with-doc}
```

File: packages/gitlab/gitlab.0.1.0/opam (hunk 2 above)
Comment (mseri) [inline packages/gitlab/gitlab.0.1.0/opam:23]:
```suggestion
  "mdx" {with-test}
  "odoc" {with-doc}
```

**Comment by tmcgilchrist** [comment]
> This seems to appear as error in all packages:
> 
> ```
> # File "cli/dune", line 7, characters 0-37:
> # 7 | (mdx
> # 8 |  (files lab.md)
> # 9 |  (packages lab))
> # Error: Program ocaml-mdx not found in the tree or in PATH
> #  (context: default)
> # Hint: opam install mdx
> # File "test/dune", line 28, characters 0-79:
> # 28 | (rule
> # 29 |  (with-stdout-to
> # 30 |   user_short_output.json
> # 31 |   (run ./diff.exe user_short)))
> # Error: No rule found for test/diff.exe
> ```
> 
> Also there are some failures related to attempting network connections (which are not allowed in the sandbox). I would suggest to disable the tests in that case

I'll do that, I need a better option for running integration tests that need network access and/or disabling them for opam repo prs.

**Comment by mseri** [comment]
We can manually disable the tests on the PR on release, it's not a problem.
opam only uses the metadata from the opam-repository, you can keep them on in your repo opam files iif you want

**Comment by tmcgilchrist** [comment]
@mseri could we disable the tests for now.

File: packages/gitlab-jsoo/gitlab-jsoo.0.1.0/opam [hunk 10]
```diff
@@ -0,0 +1,48 @@
...
+  [
+    "dune"
+    "build"
+    "-p"
+    name
+    "-j"
+    jobs
+    "--promote-install-files=false"
+    "@install"
+    "@runtest" {with-test}
```
Comment (tmcgilchrist) [inline packages/gitlab-jsoo/gitlab-jsoo.0.1.0/opam:34]:
```suggestion
```

File: packages/gitlab-unix/gitlab-unix.0.1.0/opam [hunk 11]
```diff
@@ -0,0 +1,53 @@
...
+  [
+    "dune"
+    "build"
+    "-p"
+    name
+    "-j"
+    jobs
+    "--promote-install-files=false"
+    "@install"
+    "@runtest" {with-test}
```
Comment (tmcgilchrist) [inline packages/gitlab-unix/gitlab-unix.0.1.0/opam:39]:
```suggestion
```

File: packages/gitlab/gitlab.0.1.0/opam [hunk 12]
```diff
@@ -0,0 +1,51 @@
...
+  [
+    "dune"
+    "build"
+    "-p"
+    name
+    "-j"
+    jobs
+    "--promote-install-files=false"
+    "@install"
+    "@runtest" {with-test}
```
Comment (tmcgilchrist) [inline packages/gitlab/gitlab.0.1.0/opam:37]:
```suggestion
```

File: packages/lab/lab.0.1.0/opam [hunk 13]
```diff
@@ -0,0 +1,45 @@
...
+  [
+    "dune"
+    "build"
+    "-p"
+    name
+    "-j"
+    jobs
+    "--promote-install-files=false"
+    "@install"
+    "@runtest" {with-test}
```
Comment (tmcgilchrist) [inline packages/lab/lab.0.1.0/opam:31]:
```suggestion
```

**Review by tmcgilchrist — COMMENTED** [review]
If I remove these lines will it disable running tests? @kit-ty-kate @mseri

**Comment by kit-ty-kate** [comment]
Thanks!

**Pull request merged**